### PR TITLE
[WIP] Update ststm32 to 14.x, variants to stm32duino 2.x

### PR DESCRIPTION
**Pull request by slowbro**
### Description

With the need for `utility/stm32_eeprom.h` from de0fc08, we'll need to upgrade to ststm32 14.0. That includes stm32duino 2.0.0... which changes a lot of things in variants.

### Requirements

Any STM32 board.

### Benefits

~~Fixes~~ Updates all the things!

### Configurations

Any example config for STM32.

### Related Issues

None, *yet!*

### CODE REVIEW

**Comment by slowbro** [comment]
More or less, what needs to be done for any given variant is..

1. Look at `buildroot/share/PlatformIO/boards/` in the board file for the variant. Identify the exact MCU.
2. Find the new variant files: https://github.com/stm32duino/Arduino_Core_STM32/tree/master/variants
3. In `buildroot/share/PlatformIO/variants/NAME_OF_VARIANT`:
  a. Rename variant.cpp and variant.h to variant_NAME_OF_VARIANT.cpp and variant_NAME_OF_VARIANT.h
  b. Update PinNamesVar.h from stm32duino (straight copy should be OK)
  c. Diff `variant_generic` and `variant_NAME_OF_VARIANT` cpp and h files. Apply differences, keeping Marlin's customizations..
  d. (??) Copy PeripheralPins.c over - not sure if this can be directly overwritten, or if some special stuff is needed...

File: ini/stm32f4.ini [hunk 1]
```diff
@@ -423,12 +419,10 @@ build_flags       = ${stm_flash_drive.build_flags} ${stm32f4_I2C1.build_flags}
 
 #
 # MKS Robin Nano V3 with USB Flash Drive Support and Shared Media
-# Currently, using a STM32duino fork, until USB Host and USB Device MSC get merged
 #
 [env:mks_robin_nano_v3_usb_flash_drive_msc]
 platform          = ${common_stm32.platform}
 extends           = env:mks_robin_nano_v3
-platform_packages = framework-arduinoststm32@https://github.com/rhapsodyv/Arduino_Core_STM32/archive/usb-host-msc-cdc-msc-3.zip
```
Comment (ghost) [inline ini/stm32f4.ini:431]: that's a regression as https://github.com/rhapsodyv/Arduino_Core_STM32/commit/21f67b2c3928eb9e046b79e8ff701a21dc337383 still wasn't merged in arduino_core_STM32 and we will still need a fork and a rebase on top of 2.0.0

Comment (slowbro) [reply ini/stm32f4.ini:431]: Ah, I incorrectly assumed 14.0 had those features.

File: Marlin/src/HAL/STM32/HAL.h [hunk 2]
```diff
@@ -123,7 +123,7 @@
 // Types
 // ------------------------
 
-typedef int16_t pin_t;
+typedef uint32_t pin_t;
```
Comment (thinkyhead) [inline Marlin/src/HAL/STM32/HAL.h:126]: Is it really a 32-bit integer now? That seems excessive, and might make the code a bit larger and slower than it needs to be.

**Comment by thinkyhead** [comment]
Based on the changes in this PR I get the feeling that some of our local variants might be able to go away and we could just use the framework variant for a few cases.

File: Marlin/src/HAL/STM32/HAL.h (hunk 2 above)
Comment (slowbro) [reply Marlin/src/HAL/STM32/HAL.h:126]: It's not *explicitly* defined, but I was getting overflow conversion errors from PinName to pin_t

https://github.com/stm32duino/Arduino_Core_STM32/blob/master/cores/arduino/stm32/PinNames.h

I think it has to do with the splitting on digital and analog pins. They went with `uint32_t` for analog for some unknown reason...

https://github.com/stm32duino/Arduino_Core_STM32/blob/47097af5039c90ff634cd7beb040b5b1f7074c14/variants/STM32F4xx/F446R(C-E)T/variant_generic.cpp#L71

**Comment by slowbro** [comment]
> Based on the changes in this PR I get the feeling that some of our local variants might be able to go away and we could just use the framework variant for a few cases.

Indeed. The stm32duino lib has many of the boards as variants already, i.e. FYSETC S6. Can we simply use those?

**Comment by thinkyhead** [comment]
> Can we simply use those?

Sometimes. We might have customized some, and would have to check that.

**Comment by slowbro** [comment]
Still workin' on this, trying to wrap my head around all the changes & what generic boards are available, and which custom boards we *really* need to keep around. Lots of experimenting but not much to update.

**Comment by rhapsodyv** [comment]
I started this work myself, but stopped until the next ststm32 come out for one reason: PIO didn’t like the new path scheme for the variants. The new path scheme uses ( ) , and pio said that it are breaking some tools.

There’s a discussion about that in the stm32duino repository. And I send this info to our dev channel too. 

stm32duino added some changes I asked for them, that would make use able to use generic variants, without need to copy it to marlin.

The major issue here is that we *can’t* to it all-in-one, because it would be impossible to test everything, even if you have dozens of boards at your disposal.

My suggestion is to update one env that can be fully tested, and then let users test it us for a while.

I even talked about this change with @thinkyhead , that it would generate a major release.

**Comment by rhapsodyv** [comment]
> With the need for utility/stm32_eeprom.h from de0fc08, we'll need to upgrade to ststm32 14.0. That includes stm32duino 2.0.0... which changes a lot of things in variants.

We kept a compatibility code until we are able to update everything.

**Comment by rhapsodyv** [comment]
We need to write down on top of each copied variant the reason for the copy. Without this, everything is guessing.

**Comment by slowbro** [comment]
Most of the variants are simply selecting pin functions and HAL modules.
From what I have seen thus far it can be collapsed into much fewer boards
and variants. Additionally we can get rid of the ldscript (since custom
ones are stored elsewhere), and the PinNames header file as they would
already exist in the destination dirs and are (as far as I have seen) not
modified by us.

I definitely agree that this will need a monumental amount of testing.
Breaking it into smaller chunks would be wise. I haven't seen any issues
with the parens in the names so far, but it doesn't surprise me to hear
that.

**Comment by rhapsodyv** [comment]
We probably can delete all variants copy from marlin now that pins are compile constant (before that were runtime constant, that breaks marlin fastio) and allow timers overwrite.

We just need move the timer changes for build flags, and check if the pin peripheral arrays match with the original. This is the main attention point that leads to random and hard to track bugs.

There’re others subtle changes in the marlin variants copy that can simply don’t boot the board (like changing the default HSE VALUE, or default PLL parameters).
Some ld script silly change some ram size because of the bootloader too. But I think we don’t have any of this in the stm32 Hal now.

This is the main attention point.

if I remember right, only one or two recent BTT boards doesn’t have the variant in the stm32duino yet. But I don’t check it for some weeks.

Other point is about the parêntesis. Loot at stm32duino discussions, for a thread of pio complaining about it. So, maybe it can break a debugger, or a windows compiler, or a Linux one. It can break some of python scripts too.

For the last, we need keep a policy that we only have copied variants if we write down why. 

Well.. I think that is it.

**Comment by rhapsodyv** [comment]
Ahh, for the last: all recent USB stuff (msc, usb host, etc) points to my stm32duino fork, because they didn’t merge my PR yet. So, all those envs can’t use the new stm32duino yet…
Until we update all the usb PRs too.

**Comment by rhapsodyv** [comment]
Here is the PIO complains about (): https://github.com/stm32duino/Arduino_Core_STM32/discussions/1411

I was waiting for a 15 version soon, because the way that Ivan points it, it seems something is very broken.

**Comment by slowbro** [comment]
After talking with @rhapsodyv, we determined it's best to split this into multiple updates instead of one large one due to the huge amount of testing and number of boards. Likely, it will be best to split it by the 'Fnxx' types, i.e. F103, F401, F407, F446...

The first one I'd like to tackle is F407, as I have a number of the boards that would be affected (so therefore I can test):

        - STM32F407xx
            - VxT6 (marlin_F4x7Vx): 'F407V(E-G)T_F417V(E-G)T'                  = ...
                - 407VET6:                                                     = genericSTM32F407VET6
                    - mks_robin_pro2
                    - black_f407ve (upstream)
                    - armed_v1 (upstream)
                - 407VGT6:                                                     = genericSTM32F407VGT6
                    - Anet_ET4_OpenBLT
                    - E3 RRF
                    - btt002
                    - skr 2
                    - mks_robin_nano_v3
            - ZGT6: 'F407Z(E-G)T_F417Z(E-G)T'                                   = needs new marlin_board
                - FLYF407ZG
                - marlin_BigTree_SKR_Pro
                - LERDGE S, K, X
            - IGT6: marlin_BigTree_GTR_v1 'F407I(E-G)(H-T)_F417I(E-G)(H-T)'     = genericSTM32F407IGT6

We can use upstream generic variants for most of these, but there is not one for `ZGT6` so I'd be creating/using a `marlin_STM32F407ZG` board. 

The other concern is the parenthesis. @rhapsodyv from the link you posted, it seems like stm32duino isn't super keen on actually changing/fixing that. Do you think they will do away with the parens? We can pre-process/patch the lib if absolutely necessary (i.e. `fix_framework_weakness.py`) by renaming directories/patching board files to remove parens.

We can also just table this for the time being and come back to it once 15.0 is released, especially if we're still waiting on the MSC patches to be accepted. Let me know how y'all feel about it.

**Comment by rhapsodyv** [comment]
> The first one I'd like to tackle is F407, as I have a number of the boards that would be affected (so therefore I can test):
> 

It's a good approach, but remember you will need to duplicate the envs (setting the platform only for the new envs) and keep the olds too, at least until I have time to update my stm32duino USB fork. Almost all F4 boards support usb host or msc, that depends of my stm32duino fork and it's not updated to 2.0.0 yet.


> We can use upstream generic variants for most of these, but there is not one for `ZGT6` so I'd be creating/using a `marlin_STM32F407ZG` board.

Maybe you can submit the variant to stm32duino too. What do you thing?

> 
> The other concern is the parenthesis. @rhapsodyv from the link you posted, it seems like stm32duino isn't super keen on actually changing/fixing that. Do you think they will do away with the parens? We can pre-process/patch the lib if absolutely necessary (i.e. `fix_framework_weakness.py`) by renaming directories/patching board files to remove parens.
> 
> We can also just table this for the time being and come back to it once 15.0 is released, especially if we're still waiting on the MSC patches to be accepted. Let me know how y'all feel about it.

The issue with parenthesis are that we don't know how many hidden bugs there're in PIO. So, if we change everything and PIO has a bug for windows 8 (for example), those users will need to wait until a new PIO release to be able to build Marlin. It's too risk. That is the reason I delayed it, wait to it become more stable.

But if we update it per env, without removing the old code, I'm ok with that.

This is probably one of the biggest changes  (because it basically change the ground of it) in Marlin STM32 for the lasts months.

**Comment by slowbro** [comment]
> remember you will need to duplicate the envs (setting the platform only for the new envs) and keep the olds too, at least until I have time to update my stm32duino USB fork.

Understood; I'll make them completely new/separate envs that we can use for testing until we're ready to cut over at some point.

> Maybe you can submit the variant to stm32duino too. What do you thing?

That's a good idea, especially for a simple generic board for ZGT6.

> But if we update it per env, without removing the old code, I'm ok with that.

Sounds good! I'll make sure the old envs/code is not affected. I'm still waiting on a couple of the F407 boards to arrive but will get to work on the code side, splitting things up.

**Comment by slowbro** [comment]
Going to close this now, and open a new PR for the first batch once I am finished.

**Pull request closed without merging**